go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,482 @@
+activity interceptors.
+
+Why: Maps to what Java does.
+
+### WorkerInterceptor.InterceptWorkflow
+
+Current:
+
+```go
+InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+```
+
+Proposal:
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+```
+
+Why: Callers can use the context to get workflow info in addition to other things (e.g. loggers, metricsm, etc). Also
+using the context as the first parameter is good Go practice and allows for more contextual information to be provided
+in the future.
+
+### WorkflowXInterceptor
+
+Proposal: Change the name from `WorkflowXCallsInterceptor` to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more
+accepted in Java, not Go).
+
+### WorkflowInboundInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
+```
+
+Why:
+
+* Workflow type can be derived from the context _or_ the info first given to `InterceptWorkflow`
+  * We are currently not allowing workflow type to be changed here, but it has been mentioned and we can add that to
+    `WorkflowInput` at a later time if necessary
+* Workflows can only return a value + error or just an error and this makes it more explicit what the error is
+
+### WorkflowInboundInterceptor.HandleSignal
+
+Current:
+
+```go
+ProcessSignal(ctx Context, signalName string, arg interface{}) error
+```
+
+Proposal:
+
+```go
+type SignalInput struct {
+  SignalName string
+  Arg interface{}
+}
+
+HandleSignal(ctx Context, input *SignalInput) error
+```
+
+Also, this will be called to handle the signal with the last invocation in the chain passing the signal to the workflow.
+
+Why:
+
+* `HandleSignal` matches Java nomenclature and new behavior better
+* Currently the sending of the signal to the handler is done after the interceptor not allowing the interceptor to
+  actually intercept and alter/prevent the send. We will change to have the last interceptor do the actual send.
+* We will need an integration test to confirm that this is invoked on start-with-signal workflow that has a sync receive
+  as the first statement
+
+### WorfkflowInboundInterceptor.HandleQuery
+
+Current:
+
+```go
+HandleQuery(ctx Context, queryType string, args *commonpb.Payloads,
+  handler func(*commonpb.Payloads) (*commonpb.Payloads, error)) (*commonpb.Payloads, error)
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
```

Comment:
Would this still allow an interceptor to access and potentially modify the handler's return value before bubbling it up?